### PR TITLE
flutter_tool: add and use WebCompilerConfig

### PR DESCRIPTION
**Pull request by kevmoo**
includes subclasses for JS and Wasm

Centralizes logic around configuring and applying compiler configuration

Migrated all JS logic to (new) JsCompilerConfig

Sets the stage for https://github.com/flutter/flutter/issues/124177

### CODE REVIEW

File: packages/flutter_tools/lib/src/web/compiler_config.dart [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+  /// Returns `true` if `this` represents configuration for the Wasm compiler.
+  ///
+  /// Otherwise, `false`–represents the JavaScript compiler.
+  bool get isWasm;
+
+  Map<String, String> toBuildSystemEnvironment();
+}
+
+/// Configuration for the Javascript compiler.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compiler_config.dart:16]: Throughout this file, would it make sense to substitute "dart2js" for "The JavaScript compiler", as it is more specific?

File: packages/flutter_tools/lib/src/web/compiler_config.dart [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  @override
+  Map<String, String> toBuildSystemEnvironment() => <String, String>{
+        kCspMode: csp.toString(),
+        kDart2jsDumpInfo: dumpInfo.toString(),
+        kNativeNullAssertions: nativeNullAssertions.toString(),
+        kDart2jsNoFrequencyBasedMinification: noFrequencyBasedMinification.toString(),
+        kDart2jsOptimization: optimizationLevel,
+        kSourceMapsEnabled: sourceMaps.toString(),
+      };
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compiler_config.dart:108]:
```suggestion
  Map<String, String> toBuildSystemEnvironment() => <String, String>{
    kCspMode: csp.toString(),
    kDart2jsDumpInfo: dumpInfo.toString(),
    kNativeNullAssertions: nativeNullAssertions.toString(),
    kDart2jsNoFrequencyBasedMinification: noFrequencyBasedMinification.toString(),
    kDart2jsOptimization: optimizationLevel,
    kSourceMapsEnabled: sourceMaps.toString(),
  };
```

File: packages/flutter_tools/lib/src/web/compiler_config.dart [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  /// dart2js arguments to use in both phases: full compile and CFE-only.
+  List<String> toSharedCommandOptions() => <String>[
+        if (nativeNullAssertions) '--native-null-assertions',
+        if (!sourceMaps) '--no-source-maps',
+      ];
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compiler_config.dart:114]:
```suggestion
  List<String> toSharedCommandOptions() => <String>[
    if (nativeNullAssertions) '--native-null-assertions',
    if (!sourceMaps) '--no-source-maps',
  ];
```

File: packages/flutter_tools/lib/src/web/compiler_config.dart (hunk 2 above)
Comment (kevmoo) [reply packages/flutter_tools/lib/src/web/compiler_config.dart:108]: Let's be thoughtful about where we fight dartfmt, right?

File: packages/flutter_tools/lib/src/web/compiler_config.dart [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  /// dart2js arguments to use in the full compile, but not CFE-only.
+  ///
+  /// Includes the contents of [toSharedCommandOptions].
+  List<String> toCommandOptions() => <String>[
+        ...toSharedCommandOptions(),
+        '-$optimizationLevel',
+        if (dumpInfo) '--dump-info',
+        if (noFrequencyBasedMinification) '--no-frequency-based-minification',
+        if (csp) '--csp',
+      ];
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compiler_config.dart:125]:
```suggestion
  List<String> toCommandOptions() => <String>[
    ...toSharedCommandOptions(),
    '-$optimizationLevel',
    if (dumpInfo) '--dump-info',
    if (noFrequencyBasedMinification) '--no-frequency-based-minification',
    if (csp) '--csp',
  ];
```

Comment (kevmoo) [reply packages/flutter_tools/lib/src/web/compiler_config.dart:125]: Seems so silly to spend human brain power on whitespace...

File: packages/flutter_tools/lib/src/web/compiler_config.dart [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const JsCompilerConfig({
+    required this.csp,
+    required this.dumpInfo,
+    required this.nativeNullAssertions,
+    required this.optimizationLevel,
+    required this.noFrequencyBasedMinification,
+    required this.sourceMaps,
+  });
+
+  const JsCompilerConfig.run({required bool nativeNullAssertions})
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compiler_config.dart:27]: is this a constructor? i'm also unfamiliar with the `this()` syntax, is this syntactic sugar over `this.csp = false, this.dumpInfo = false, ...`?

Comment (kevmoo) [reply packages/flutter_tools/lib/src/web/compiler_config.dart:27]: Yes and yes.

File: packages/flutter_tools/lib/src/commands/build_web.dart [hunk 6]
```diff
@@ -133,9 +133,21 @@ class BuildWebCommand extends BuildSubCommand {
       throwToolExit('"build web" is not currently supported. To enable, run "flutter config --enable-web".');
     }
 
-    final bool wasmRequested = boolArg('wasm');
-    if (wasmRequested && !featureFlags.isFlutterWebWasmEnabled) {
-      throwToolExit('Compiling to WebAssembly (wasm) is only available on the master channel.');
+    WebCompilerConfig compilerConfig;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/build_web.dart:136]:
```suggestion
    final WebCompilerConfig compilerConfig;
```

File: packages/flutter_tools/lib/src/web/compiler_config.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/web/compiler_config.dart:27]: can you add a dartdoc describing when to call this constructor (it's not obvious from the name)

**Review by christopherfujino — APPROVED** [review]
some nits, but LGTM

**Pull request merged**